Add QEMU support to opentitanlib

jwnrt opened the pull request:
## Summary

This PR adds support for the QEMU emulator to opentitanlib.

This is based on top of https://github.com/lowRISC/opentitan/pull/19420 which is why its commits are here too.

## Testing

You can run some tests on QEMU using the `_qemu_cw310_` target, e.g:

```sh
bazel test --test_output=streamed --cache_test_results=no //sw/device/tests/crypto:aes_functest_qemu_cw310_rom_with_fake_keys
bazel test --test_output=streamed --cache_test_results=no //sw/device/tests:aes_idle_test_qemu_cw310_test_rom
```

## Status

This PR currently only supports interacting with the UART. It extracts Verilator's file-based UART connection to the `common::uart` module and reuses that.

## Future work

The following will be implemented in future PRs:

- SPI
- GPIO (not fully implemented for the ot-earlgrey QEMU machine)
- Flash and OTP writing (not fully implemented for the ot-earlgrey QEMU machine, requires us to make a writable copy of the readonly image files given to us by Bazel)

pamaury commented on sw/host/opentitanlib/src/app/config/opentitan_qemu.json:
```diff
@@ -0,0 +1,46 @@
+{
+  "includes": ["/__builtin__/opentitan.json"],
+  "interface": "qemu",
+  "pins": [
```
I am guessing that those definitions are just to make opentitantool happy and that QEMU OT does not support gpios yet?

jwnrt replied:
That's what I added them for, although now I see that I can remove them

pamaury commented on sw/host/opentitanlib/src/transport/qemu/subprocess.rs:
```diff
@@ -0,0 +1,235 @@
+    /// The ROM image used to boot the CPU.
+    pub rom_image: Option<PathBuf>,
+
+    /// The flash images stored in internal flash memory, one file per bank.
```
I think there is just one image for all data banks

pamaury commented on sw/host/opentitanlib/src/transport/qemu/subprocess.rs:
```diff
@@ -0,0 +1,235 @@
+        // Start QEMU.
+        let mut child = command
+            .stdin(Stdio::null())
+            .stdout(Stdio::piped())
```
What happens to the stdout? Shouldn't it be either piped and printed, or nulled?

jwnrt replied:
I'll null it since I don't expect to use QEMU's stdout for any of the sockets

jwnrt replied:
Although now I see that one of the unit tests is wanting to read from `stdout` so maybe I'll accumulate it in the same place as the stderr.

pamaury commented on sw/host/opentitanlib/src/transport/qemu/subprocess.rs:
```diff
@@ -0,0 +1,235 @@
+
+    /// Finds a string within the QEMU output.
+    /// It is assumed that the [`Regex`] `re` has zero or one capture groups.
+    pub fn find(&self, re: &Regex, deadline: Instant) -> anyhow::Result<String> {
```
I am struggling to see the flow here: this function is called after creating the subprocess. But the subprocess constructor already spawns a thread to print the accumulated stderr. Isn't there a risk that the printer will eat the message you are looking for before this function has the time to lock `self.accumulated_stderr`?

jwnrt replied:
I think the `printer` is just redirecting `stderr` to the `Subprocess::accumulated_stderr` string which is what `fn find` is reading from, so they aren't both trying to read from the same source.

I'm not very happy with this function but copied it from the verilator module (maybe should have moved it somewhere to share)

pamaury commented on sw/host/opentitanlib/src/transport/qemu/transport.rs:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    fn dispatch(&self, action: &dyn Any) -> Result<Option<Box<dyn Annotate>>> {
+        if let Some(watch) = action.downcast_ref::<Watch>() {
```
Is `Watch` used anywhere?

jwnrt replied:
I don't understand how `Watch` works, but it's used by `fn dispatch`. I copied this from the verilator module but should have probably tried to understand it first.

jwnrt replied:
Deleting `fn dispatch()` and `Watch` doesn't cause any issues, so maybe they're just fine to leave?

The documentation for `dispatch` says:

```rust
/// Invoke non-standard functionality of some Transport implementations.
```

which doesn't help much

jwnrt commented:
I've rebased this PR on top of @pamaury's latest QEMU support in Bazel, and also changed the UART implementation to use the exist `SerialPortUart` struct that the CW310 is using.

This required changing the output device QEMU's using to a pty (or tty). Unfortunately, we've hit some issues where QEMU will blackhole any data sent to a PTY if there's nothing connected to the other side.

This means we have to halt the CPU on startup, wait until we've connected the PTY, then send a command to start the CPU.

For some reason, we have to add a one-second delay between connecting to the PTY and sending the start command. I'm not sure why this is, but a 750ms delay wasn't enough. Maybe QEMU takes some time to recognise that there's something connected to the other end. I can't find any reliable way of checking if QEMU knows there's something connected.

Not sure what the proper solution to this issue is, but we may have to revert back to using the manual file-based UART implementation.

pamaury commented:
We found the cause of the delay with @jwnrt: the pty code in qemu only polls for the connected state [on read or every second](https://github.com/lowRISC/qemu/blob/ot-earlgrey-8.0.2/chardev/char-pty.c#L170). Since the uart is only written to, it means that it can take up to one second for QEMU to realize that we have connected.

I will look into patching the QEMU source code if possible to avoid this delay.

pamaury commented on sw/host/opentitanlib/src/backend/qemu.rs:
```diff
@@ -0,0 +1,63 @@
+    /// Name of the QEMU machine to execute.
+    #[arg(long, default_value = "ot-earlgrey")]
+    pub qemu_machine: String,
+
```
```suggestion
Additional properties of the QEMU machine.
```

pamaury commented on sw/host/opentitanlib/src/backend/qemu.rs:
```diff
@@ -0,0 +1,63 @@
+    #[arg(long)]
+    pub qemu_rom: Option<PathBuf>,
+
+    /// Paths to flash images to use.
```
```suggestion
    /// Path to the flash image to use.
```

pamaury commented on sw/host/opentitanlib/src/transport/qemu/mod.rs:
```diff
@@ -0,0 +1,11 @@
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+//! This transport spawns a QEMU subprocess and connects to its IO via sockets.
```
```suggestion
//! This transport spawns a QEMU subprocess and connects to its IO via the most appropriate method.
```

pamaury replied:
Or maybe just say it connects to its IO.

jwnrt commented on sw/host/opentitanlib/src/transport/common/uart.rs:
```diff
@@ -273,3 +273,142 @@ pub fn flock_serial(port: &TTYPort, port_name: &str) -> Result<()> {
     Ok(())
 }
+
+/// Represents a file-backed UART connection.
```
That's what [this commit](https://github.com/lowRISC/opentitan/pull/19471/commits/beb1d74eebb5bd3a2dd00a470cf3e69952a769b5) should be, is that okay?

rivos-eblot commented:
> The latest QEMU flashgen changes fixed most errors, but there are still some remaining:

Should be addressed with https://github.com/lowRISC/qemu/commit/dc9c08d8acfda5170995ff9fee6494090175ef50

jwnrt commented:
I've added two commits for trying out the flashgen -A flag, but these should probably:

1. be moved into https://github.com/lowRISC/opentitan/pull/19420
2. only be applied to known-bad inputs and not globally

jwnrt commented:
Rebased

jwnrt commented:
It wasn't superceded but it's no longer useful, we would redo it from scratch
